### PR TITLE
Add cumsum

### PR DESCRIPTION
**Pull request by Illviljan**

### CODE REVIEW

File: flox/aggregations.py [hunk 1]
```diff
@@ -470,6 +470,19 @@ def _pick_second(*x):
 # median = Aggregation("median", chunk=None, combine=None, fill_value=None)
 # nanmedian = Aggregation("nanmedian", chunk=None, combine=None, fill_value=None)
 
+# Cumulatives:
+cumsum_ = Aggregation(
```
Comment (dcherian) [inline flox/aggregations.py:474]: Let's rename `reduction_type` to `kind` and set it to `cumulative` here.

File: flox/aggregations.py [hunk 2]
```diff
@@ -470,6 +470,19 @@ def _pick_second(*x):
...
+    "cumsum",
+    chunk="cumsum",
+    combine="sum",
+    fill_value=0,
```
Comment (dcherian) [inline flox/aggregations.py:478]: fill_value and final_fill_value don't amke any sense?

Set `chunk=None` to raise an error with dask

File: flox/core.py [hunk 3]
```diff
@@ -743,7 +743,11 @@ def chunk_reduce(
 
     assert by.ndim <= array.ndim
 
-    final_array_shape = array.shape[:-nax] + (1,) * (nax - 1)
```
Comment (dcherian) [inline flox/core.py:746]: I'd prefer a `chunk_accumulate` instead. Hopefully there won't be too much duplication.

File: tests/test_core.py [hunk 4]
```diff
@@ -1463,3 +1463,15 @@ def test_method_check_numpy():
...
+
+
+def test_cumsum() -> None:
+    import numpy_groupies as npg
+
+    group_idx = np.array([4, 3, 3, 4, 4, 1, 1, 1, 7, 8, 7, 4, 3, 3, 1, 1])
+    a = np.array([3, 4, 1, 3, 9, 9, 6, 7, 7, 0, 8, 2, 1, 8, 9, 8])
+    b = npg.aggregate(group_idx, a, func="cumsum")
+
+    bb = groupby_reduce(a, group_idx, func="cumsum", engine="numpy")[0]
```
Comment (dcherian) [inline tests/test_core.py:1475]: Let's add a `groupby_accumulate(a, group_idx, func="sum", engine="numpy)` that only returns a single array.

**Review by dcherian — COMMENTED** [review]
Exciting to see!

File: flox/core.py (hunk 3 above)
Comment (Illviljan) [reply flox/core.py:746]: chunk reduce is used every! 
I'll try out something like:
```
def chunk_aggregate():
  if is_accumulate:
    chunk_accumulate
    ....
```

Comment (Illviljan) [reply flox/core.py]: I think I got what you meant now. Now with lots of copy/pasting

File: tests/test_core.py (hunk 4 above)
Comment (Illviljan) [reply tests/test_core.py]: I'm not a fan that groupby_reduce and groupby_accumulate wont have the same typing though.

File: flox/aggregations.py (hunk 2 above)
Comment (Illviljan) [reply flox/aggregations.py]: done

File: flox/aggregations.py (hunk 1 above)
Comment (Illviljan) [reply flox/aggregations.py:474]: Done

File: tests/test_core.py (hunk 4 above)
Comment (dcherian) [reply tests/test_core.py]: They are very different operations so they have to be. For e.g. `method="map-reduce"` isn't applicable.

**Comment by Illviljan** [comment]
There's no implementation of `cumprod` in` npg.aggregate_numpy`. So that won't quickly just work.
https://github.com/ml31415/numpy-groupies/blob/9fc51ad22843ac9e5d3c65ae8444e74ae0a71148/numpy_groupies/aggregate_numpy.py#L234-L261

File: flox/core.py [hunk 5]
```diff
@@ -2052,3 +2057,562 @@ def groupby_reduce(
...
+    sort: bool = True,
+    isbin: T_IsBins = False,
+    axis: T_AxesOpt = None,
+    fill_value=None,
+    dtype: np.typing.DTypeLike = None,
+    min_count: int | None = None,
+    method: T_Method = "map-reduce",
+    engine: T_Engine = "numpy",
+    reindex: bool | None = None,
+    finalize_kwargs: dict[Any, Any] | None = None,
```
Comment (dcherian) [inline flox/core.py:2303]:
```suggestion
    func: T_Agg,
    expected_groups: T_ExpectedGroupsOpt = None, #TODO: think about this one
    isbin: T_IsBins = False,
    axis: T_AxesOpt = None,
    dtype: np.typing.DTypeLike = None,
    engine: T_Engine = "numpy",
```

File: flox/core.py [hunk 6]
```diff
@@ -2052,3 +2057,562 @@ def groupby_reduce(
...
+                groups = (groups[0][sorted_idx],)
+
+    if factorize_early:
+        # nan group labels are factorized to -1, and preserved
+        # now we get rid of them by reindexing
+        # This also handles bins with no data
+        result = reindex_(result, from_=groups[0], to=expected_groups, fill_value=fill_value)
+        groups = final_groups
+
+    return (result, *groups)  # type: ignore[return-value]  # Unpack not in mypy yet
```
Comment (dcherian) [inline flox/core.py:2565]:
```suggestion
    return result
```

File: flox/core.py [hunk 7]
```diff
@@ -2052,3 +2057,477 @@ def groupby_reduce(
...
+        # now we get rid of them by reindexing
+        # This also handles bins with no data
+        result = reindex_(result, from_=groups[0], to=expected_groups, fill_value=_fill_value)
+        groups = final_groups
+
+    return result
+
+
+# %% Aggregate
+def groupby_aggregate(
```
[the remark this answers is not in the document]
Comment (Illviljan) [reply flox/core.py]: done

File: flox/aggregations.py [hunk 8]
```diff
@@ -496,6 +513,9 @@ def _pick_second(*x):
     "nanfirst": nanfirst,
     "last": last,
     "nanlast": nanlast,
+    # Cumulatives:
+    "cumsum": cumsum_,
+    "cumprod": cumprod_,
```
Comment (Illviljan) [inline flox/aggregations.py:518]:
```suggestion
```

File: flox/aggregations.py [hunk 9]
```diff
@@ -470,7 +471,23 @@ def _pick_second(*x):
...
+    chunk=None,
+    combine="sum",
+    kind="cumulate",
+)
+cumprod_ = Aggregation(
+    "cumprod",
+    chunk=None,
+    combine="prod",
+    kind="cumulate",
+)
```
Comment (Illviljan) [inline flox/aggregations.py:486]:
```suggestion
```

**Pull request closed without merging**